### ISSUE TITLE
Keyboard navigation with pagination - stuck on a single page.

### ISSUE THREAD

**Issue by Ignasimg** [ag-grid/ag-grid#4426]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => see 'Providing a Reproducible Scenario'
[] feature request => do not use Github for feature requests, see 'Customers of AG Grid'
[] support request => see 'Requesting Community Support'
```

**Current behavior**
When `pagination: true` keyboard navigation is stuck on same page.
Arrow down never goes beyond the last result of the page; Nor pageDown which jumps to the last entry but not to the next page. (Similar for arrow up and pageUp)

Sample:
https://plnkr.co/edit/mqgiHilw19PNuY1s

**Expected behavior**
I'd expect to be able to navigate to the next page with the keyboard.

* **AG Grid version:** 25.1.0

* **Browser:**
Chrome, Firefox

**Comment by makinggoodsoftware** [comment]
Hi,

Thanks for your feedback, note that we don’t think this is a bug, as the system works as expected in this area.

In our opinion, this would be a feature request

Note that we don’t have the capacity to take feature requests as we have a full pipeline.
https://www.ag-grid.com/ag-grid-pipeline/

If you are a customer, you should be using ag-Grid Enterprise in which case you should have access to our customer support system (via zendesk), and that is the only channel where we accept feature requests.

If you are an ag-Grid Enterprise Customer, please raise this in our Zendesk support system - contact info@ag-grid.com for access. This is our primary channel for Support.

---

We are going to label this as managed-by-the-community as it is probably there are some ways to mitigate this and hopefully someone from the community can help you with a workaround/fix.

**Comment by Ignasimg** [comment]
Hello @makinggoodsoftware 

I'd greatly appreciate if you could reconsider this; Here am going to expose why I think this is not a feature request but a conceptual bug.

* In previous versions of ag-grid it was possible to navigate between pages simply with keyboard arrows. I don't mean to say that the previous behaviour was correct nor that now is impossible (you can still tab all the way out to the pagination section and use the commands from there). But this has a big impact on accessibility.
* When pagination is active and the user is viewing the bottom of the current page, `pageDown` or `Ctrl + Arrow Down` do exactly the same (focus the last row of the current page) - this is specially ridiculous when using an option like `paginationAutoPageSize: true` my humble opinion is that this is a conceptual bug - why would `pageDown` or `pageUp` be limited to use inside the same page?

Best regards!

**Comment by makinggoodsoftware** [comment]
Hi,

We have flagged this as a good_suggestion

This means that this is a candidate to be in a future release.

At this moment we don’t have a ticket number for this that you can use to track in our pipeline:
https://www.ag-grid.com/ag-grid-pipeline/

But we should be updating this ticket with this information as soon as we process the good suggestions, as we do these in batches. Note also that at this stage we can’t guarantee that we will deliver this, but we will take it into consideration

Thanks for your feedback!

**Comment by kiril-matev** [comment]
Hello,

Thank you for contacting us about this.

Accessibility in AG Grid is implemented directly applying the WAI-ARIA standard so it conforms with the guidelines in WCAG.

Please take a look at the WAI-ARIA section concerning keyboard navigation in data grid controls:
https://www.w3.org/TR/wai-aria-practices-1.2/#keyboard-interaction-for-data-grids

Quoting from this section on the behavior of Page Up/Page Down keys:
**Page Down**: Moves focus down an author-determined number of rows, typically scrolling so the bottom row in the currently visible set of rows becomes one of the first visible rows. If focus is in the last row of the grid, focus does not move.
**Page Up**: Moves focus up an author-determined number of rows, typically scrolling so the top row in the currently visible set of rows becomes one of the last visible rows. If focus is in the first row of the grid, focus does not move.

This is why the **Page Down**, **Page Up** keys are used to navigate one page up/down in the currently displayed records on this page, and their use is restricted to the list of records displayed on the current page. This is the behavior that your users will see on any website which is compliant with the WCAG standard.

This is why we recommend using the behavior in AG Grid as is as it's compliant with accessibility standards.

If you'd like to break WCAG accessibility standard compliance anyway and allow your users to change the current page using the Page Down/Page Up keys, you can do so by using the AG Grid keyboard event, pagination and focus API as shown in this sample:
https://plnkr.co/edit/NdMttOpwl6i9NVaT

Thank you for submitting this - we appreciate your time!

**Comment by Ignasimg** [comment]
Thanks for your reply @kiril-matev 

Risking being called a nitpicker or who knows what I'll just point out two parts of the text you quote:

> Page Down: Moves focus down an author-determined number of rows, **_typically_** scrolling so the bottom row in the currently visible set of rows becomes one of the first visible rows. If focus is in **_the last row of the grid_**, focus does not move.
Page Up: Moves focus up an author-determined number of rows, **_typically_** scrolling so the top row in the currently visible set of rows becomes one of the last visible rows. If focus is in **_the first row of the grid_**, focus does not move.

1. I'm not a native english speaker but afaik typically doesn't mean always. Hence IMHO switching to prev/next page can also be a valid action.
2. The real issue with the quoted text is what do you understand by "last/first row of the grid" is the first/last row of a page the first/last row of a grid? For me it definitely is not (unless it happens to be the first row of the first page or the last row of the last page).

Having said that, I'm not here to discuss about semantics, I appreciate your time dealing with this issue, and the time you guys for surely spent discussing about it. But did you try the actual problem?

Thankfully god I'm not blind, but besides being a developer I'm a user of the aplication I build, and since for me it's annoying to pick the mouse to switch pages, or tab back and forth to get to the pagination controls I can imagine for a blind user the experience will be much much worst.

IMHO this is not about compliance with standards which we could spend days discussing about how to interpret each word not to break it ;) but about usability - and TBH if you guys did try the problem I can not believe that I'm even pointing it out.

Anyway I know this post probably got a little bit awkward but it's truly the way I feel about this matter.

Best regards and thank you for your time & patience.

**Comment by kiril-matev** [comment]
Hi @Ignasimg 

Thank you for this response and for letting me clarify my earlier comments.

Please note the quote from the standard I included above outlines the keyboard navigation behavior for data grids when using paging - using PgUp/PgDn keys navigates the records within the current page. The keyboard navigation behavior you're describing (using PgUp/PgDn keys to change the current page) applies to non-paged data grids, where all the records are visible instead of on separate pages. In such a non-paged dataset, you can use the keyboard navigation keys to go through the entire dataset. If that's the behavior you'd like to have please turn off the paging feature in AG Grid and you'll get this behavior instead.

We trust the experts of the W3C consortium who built the WCAG standard are in the best position to evaluate the usability implications of different keyboard navigation and labelling modes to suggest solutions that support assistive technology solutions such as screen readers for users with accessibility requirements. This is why it's important for us to be in compliance with accessibility standards as many of our customers perform audits of AG Grid based on these standards. 

However, if you insist on using paging and being able to change the current page by pressing PgUp/PgDn, I sent a sample which implements it earlier:
https://plnkr.co/edit/NdMttOpwl6i9NVaT

Can you please use this sample in your project?

**Comment by makinggoodsoftware** [comment]
Hi,

 

This ticket has been flagged as managed-by-community/waiting-for-repro for a while now and there has not been any activity, to help us tidy up we are going to close it.

 

Thanks
